[Enhancement] Prune the pure pushdown predicate column after filtering by GIN

srlch opened the pull request:
## Why I'm doing:
Currently, if a column is a pure pushdown predicate column with index  , which means that it satisfy:
1. It can be pushdown into storage engine and filter by index.
2. delete predicates do not contain this column.
3. output schema do not contain this column.

This means that if all the predicates of this column can be filter by index and being erased to be empty, the column need not be read at all. But in current implementation, the useless read will exist and cause the IO overhead especially for GIN

## What I'm doing:
If the column can be prune for sure, skip the page read (just resize it instead)

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

kangkaisen commented on be/src/common/config.h:
```diff
@@ -1273,4 +1273,6 @@ CONF_mInt64(jit_lru_cache_size, "0");
 CONF_mInt64(arrow_io_coalesce_read_max_distance_size, "1048576");
 CONF_mInt64(arrow_read_batch_size, "4096");
+
+CONF_mBool(enable_index_filter_column_prune, "true");
```
should use session variable.

srlch replied:
Fixed it

kangkaisen commented on fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java:
```diff
@@ -969,8 +969,8 @@ public Void visitMatchExpr(MatchExpr node, Scope scope) {
                 throw new SemanticException("left operand of MATCH must be column ref");
             }
 
-            if (!type2.isStringType() && !type2.isNull()) {
```
why must be StringLiteral?

trueeyu commented on be/src/storage/rowset/segment_iterator.cpp:
```diff
@@ -136,6 +136,22 @@ class SegmentIterator final : public ChunkIterator {
+                    //    the chunk size.
+                    // 2. In debug/ASAN mode, chunk::check_or_die will check all column have the
+                    //    same size or not.
+                    if (i == 0) {
```
So tricky, better to resize always?

trueeyu replied:
This feels too tricky. It relies on num_rows() to obtain the row count of the first column. It is easy to cause problems later. If you take out the Column and then look at the row_count, there will be problems, and it is related to the basics data structures are coupled,

srlch replied:
Fixed it, resize all columns

wyb commented on be/src/storage/rowset/segment_iterator.cpp:
```diff
@@ -136,6 +136,11 @@ class SegmentIterator final : public ChunkIterator {
                 const ColumnPtr& col = chunk->get_column_by_index(i);
+                if (_prune_column_after_index_filter && _prune_cols.count(i)) {
+                    // make sure each pruned column has the same size as the unpruneable one.
+                    col->resize(range.span_size());
```
why not directly remove the prune column from context chunk?

srlch replied:
I think it is a more safety way to implement this optimization. Remove the column in chunk need to change more code in _build_context which maybe a little aggressive and more risky

wyb commented on fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:
```diff
@@ -1135,6 +1138,9 @@ public static MaterializedViewRewriteMode parse(String str) {
     private boolean enableFilterUnusedColumnsInScanStage = true;
 
+    @VariableMgr.VarAttr(name = ENABLE_PRUNE_COLUMN_AFTER_INDEX_FILTER)
+    private boolean enablePruneColumnAfterIndexFilter = true;
```
when should it be set to false?

srlch replied:
This is an emergency session variable to turn off the optimization if the optimization may has the bug.

wyb replied:
suggest change it to invisible

srlch replied:
Done, fixed it

srlch committed:
[Enhancement] Prune the pure pushdown predicate column after filtering by GIN

Why I'm doing:
Currently, if a column is a pure pushdown predicate column, which means that it satisfy:
1. It can be pushdown into storage engine and filter by index.
2. delete predicates do not contain this column.
3. output schema do not contain this column.

This means that if all the predicates of this column can be filter by index and being erased
to be empty, the column need not be read at all. But in current implementation, the useless read
will exist and cause the IO overhead especially for GIN

What I'm doing:
If the column can be prune for sure, skip the page read (just resize it instead)

Signed-off-by: srlch <linzichao@starrocks.com>

ZiheLiu commented on be/src/storage/tablet_reader.cpp:
```diff
@@ -283,6 +283,7 @@ Status TabletReader::get_segment_iterators(const TabletReaderParams& params, std
     rs_opts.rowid_range_option = params.rowid_range_option;
     rs_opts.short_key_ranges_option = params.short_key_ranges_option;
     rs_opts.asc_hint = _is_asc_hint;
+    rs_opts.prune_column_after_index_filter = params.prune_column_after_index_filter;
```
Does lake OlapTable also support this? If so, `storage/lake/tablet_reader.cpp` and `storage/lake/rowset.cpp` also need pass this.

srlch replied:
Currently, GIN only used for OlapTable but not lake. And this optimization can only used for GIN. So lake table will not affect by this for now

ZiheLiu commented on fe/fe-core/src/main/java/com/starrocks/planner/OlapScanNode.java:
```diff
@@ -977,6 +977,8 @@ protected void toThrift(TPlanNode msg) {
             }
 
             msg.olap_scan_node.setUse_pk_index(usePkIndex);
+            msg.olap_scan_node.setEnable_prune_column_after_index_filter(
```
Does lake table also support this? If so, this also need to be add in the `if (olapTable.isCloudNativeTableOrMaterializedView())` branch.
